### PR TITLE
ARROW-10467: [FlightRPC][Java] Add the ability to pass arbitrary client headers.

### PR DESCRIPTION
**Pull request by kylepbit**
PR for https://issues.apache.org/jira/browse/ARROW-10467

Add a CallOption for specifying arbitrary properties as headers for RPC calls.
Add a ServerMiddleware that will intercept and pass the properties to a handler
when registered via the FlightServer builder.

### CODE REVIEW

**Comment by lidavidm** [comment]
What do you think about instead exposing headers as key-value metadata directly (on the sending side with a CallOption, on the receiving side via CallContext)? You could even still implement the server side like you've done here, with a standard middleware that can be configured.

**Comment by lidavidm** [comment]
Ah, I see you want some keys that wouldn't necessarily be valid headers.

**Comment by lidavidm** [comment]
Just stepping back for some context - what's the use case you have in mind for this?

**Comment by kylepbit** [comment]
The intent is to be able to provide preferences to requests that would not be part of the actual action. Ex context to run the action in, etc.

**Comment by lidavidm** [comment]
Ok. I think it'd make more sense to implement better support for headers in general, rather than have an interface to a single hardcoded header in Flight - anything higher-level (such as encoding an options map) can then be implemented in the application. Does that sound reasonable?

**Comment by kylepbit** [comment]
Yes - do you have suggestions as to how you think that would be best done? Perhaps a CallOption to specify a header and give it a key and value where the value is serializable, and have a method that serializes the value that can be overridden if people wish to specialize?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package org.apache.arrow.flight;
+
+import java.util.function.Consumer;
+
+/**
+ * Interface for server side property handling.
+ */
+public interface ServerHeaderHandler extends Consumer<CallHeaders> {
```
Comment (tifflhl) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java:25]: Should we add a unit test that demonstrate a sample usage for this handler?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderMiddleware.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.flight;
+
+/**
+ * Middleware that's used to extract and pass properties to the server during requests.
+ */
+public class ServerHeaderMiddleware implements FlightServerMiddleware {
```
Comment (tifflhl) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderMiddleware.java:23]: Can we rename ServerHeaderMiddleware to ServerPropertyMiddleware? The current name is a bit too generic.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java (hunk 1 above)
Comment (tifflhl) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java:25]: Can we rename ServerHeaderHandler to ServerPropertyHandler? Like ServerHeaderMiddleware, the current name is a too generic.

Comment (keeratsingh) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java:25]: I agree with @tifflhl's suggestion the name is too generic, would something like `ServerPropertyHeaderHandler` and `ClientPropertyHeaderHandler` be more appropriate similar to `ClientBearerHeaderHandler`.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java [hunk 3]
```diff
@@ -351,6 +359,14 @@ public Builder headerAuthenticator(CallHeaderAuthenticator headerAuthenticator)
       return this;
     }
 
+    /**
+     * Set the header handler.
+     */
+    public Builder headerHandler(ServerHeaderHandler headerHandler) {
```
Comment (keeratsingh) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:365]: Can we rename this to propertyHeaderHandler if we decide to rename `ServerHeaderHandler` 
 to `ServerPropertyHeaderHandler`?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightConstants.java [hunk 4]
```diff
@@ -23,5 +23,6 @@
 public interface FlightConstants {
 
   String SERVICE = "arrow.flight.protocol.FlightService";
+  String PROPERTY_HEADER = "Arrow-Properties";
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightConstants.java:26]: Let's rename this to something more specific (HEADERS_KEY?) and make it an instance of `FlightServerMiddleware.Key`. Let's also change the constant to something that's more clearly namespaced under arrow, e.g. "org.apache.arrow.flight.ServerHeaderHandler".

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java (hunk 1 above)
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java:25]: This interface confuses me, as it seems within an RPC handler, there's no way to get the headers associated with that particular call.

File: java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestCallOptions.java [hunk 5]
```diff
@@ -64,6 +67,61 @@ public void underTimeout() {
     });
   }
 
+  @Test
+  public void singleProperty() {
+    final MetadataAdapter headers = new MetadataAdapter(new Metadata());
```
Comment (lidavidm) [inline java/flight/flight-core/src/test/java/org/apache/arrow/flight/TestCallOptions.java:72]: We probably need a CallHeaders implementation that doesn't come from the flight-grpc package, or there's no way to use HeaderCallOption.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java (hunk 1 above)
Comment (kylepbit) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java:25]: This is not specific to properties - it's generally for handling all headers so I think the name fits.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderMiddleware.java (hunk 2 above)
Comment (kylepbit) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderMiddleware.java:23]: I think the name is appropriate given that this is intended to handle all headers for calls, which could be considered properties depending on the FlightServer implementation, but not necessarily.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java (hunk 1 above)
Comment (kylepbit) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java:25]: There already is: see https://github.com/apache/arrow/pull/8572/files#diff-b19522af7605b7cfe033ecc1300f2f6b0347d7eaba2969f0193086eb9751a546R102

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java (hunk 3 above)
Comment (kylepbit) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightServer.java:365]: As mentioned in another comment, the intent is to handle all headers. You can interpret them as properties, but I think the name is reflective of intent.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java (hunk 1 above)
Comment (kylepbit) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java:25]: Not sure I understand the comment, are you simply saying you can't correlate the headers that are incoming with the RPC call that generated them?

**Comment by lidavidm** [comment]
Just a higher level comment - I see on ARROW-10428 cookie support is being worked on and ARROW-10427 is about session headers. Since cookies are a particular header and sessions are (usually) a particular cookie, is there any interest in making sure all these features work together and are built on top of each other?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java (hunk 1 above)
Comment (lidavidm) [reply java/flight/flight-core/src/main/java/org/apache/arrow/flight/ServerHeaderHandler.java:25]: Yes - within the body of an RPC handler, how do I access the headers that the client sent for that call?

**Comment by kylepbit** [comment]
Regarding the comment about sessions and cookies - yes, there is definite interest in making sure that the work there is building on top of, instead of parallel to, one another.

**Comment by kylepbit** [comment]
@lidavidm any other comments on this?

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightCallHeaders.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+    this.keysAndValues = ArrayListMultimap.create();
+  }
+
+  @Override
+  public String get(String key) {
+    if (key.endsWith(Metadata.BINARY_HEADER_SUFFIX)) {
+      return new String((byte[]) Iterables.get(this.keysAndValues.get(key), 0));
+    }
+
+    return (String) Iterables.get(this.keysAndValues.get(key), 0);
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/FlightCallHeaders.java:47]: Iterables.get appears to throw if there's no element; for consistency with MetadataAdapter this should return null when there's no element.

File: java/flight/flight-core/src/main/java/org/apache/arrow/flight/HeaderCallOption.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+  /**
+   * Header property constructor.
+   *
+   * @param headers the headers that should be sent across. If a header is a string, it should only be valid ASCII
+   *                characters. Binary headers should end in "-bin".
+   */
+  public HeaderCallOption(CallHeaders headers) {
+    for (String key : headers.keys()) {
+      if (key.endsWith(Metadata.BINARY_HEADER_SUFFIX)) {
+        propertiesMetadata.put(
```
Comment (lidavidm) [inline java/flight/flight-core/src/main/java/org/apache/arrow/flight/HeaderCallOption.java:39]: This discards secondary values of multi-valued headers.

**Review by lidavidm — COMMENTED** [review]
Overall looks good. Two minor comments.

**Review by lidavidm — APPROVED** [review]

**Comment by lidavidm** [comment]
Thanks @kylepbit!

**Pull request closed without merging**